### PR TITLE
BLD: Add libflame as a LAPACK back-end

### PR DESCRIPTION
**Pull request by zerothi**
This adds libflame to the available list of LAPACK implementations.

It requires #13132 before merging (which is why this is a draft for now!)

### CODE REVIEW

**Comment by charris** [comment]
This will need a release note at some point.

**Comment by zerothi** [comment]
@charris could you guide me towards this? Where should I add stuff? (note I already added flame to the documentation).

**Comment by rgommers** [comment]
> Where should I add stuff? 

If you could add a note on both this PR and gh-13132 to `doc/release/1.17.0-notes.rst`, that would be very helpful. Under "new features"

**Comment by zerothi** [comment]
@rgommers could you see if 3b5b21c is fine?

File: doc/release/1.17.0-notes.rst [hunk 1]
```diff
@@ -99,8 +131,8 @@ identity, it is necessary to also pass in an initial value (e.g.,
 ``initial=np.inf`` for ``np.min``).  For instance, the equivalent of
 ``nansum`` would be, ``np.sum(a, where=~np.isnan(a))``.
 
-Timsort has replaced mergesort as the stable sorting implementation
--------------------------------------------------------------------
+Timesort has replaced mergesort as the stable sorting implementation
```
Comment (rgommers) [inline doc/release/1.17.0-notes.rst]: this change is incorrect, the name is Timsort (invented by a guy named Tim:)).

File: doc/release/1.17.0-notes.rst [hunk 2]
```diff
@@ -90,6 +90,38 @@ C API changes
...
 
+libFLAME
+--------
+Enabled the use of the libFLAME linear algebra package as a drop-in
+replacement for NetLIB LAPACK, see
+`libFLAME <https://www.cs.utexas.edu/~flame/web/libFLAME.html>`_ for details.
+
+User-defined BLAS detection order
+---------------------------------
+``np.distutils`` now uses an environment variable, comma-separated and case
```
Comment (rgommers) [inline doc/release/1.17.0-notes.rst]: nitpick: `numpy.distutils` would be better here

File: doc/release/1.17.0-notes.rst [hunk 3]
```diff
@@ -90,6 +90,38 @@ C API changes
 New Features
 ============
 
+libFLAME
+--------
+Enabled the use of the libFLAME linear algebra package as a drop-in
+replacement for NetLIB LAPACK, see
```
Comment (rgommers) [inline doc/release/1.17.0-notes.rst]: Perhaps: "Support for building NumPy with the libFLAME linear algebra package as the LAPACK implementation" ?

File: doc/source/user/building.rst [hunk 4]
```diff
@@ -155,9 +155,10 @@ The default order for the libraries are:
 
 1. MKL
 2. OpenBLAS
-3. ATLAS
-4. Accelerate (MacOS)
-5. LAPACK (NetLIB)
+3. flame
```
Comment (rgommers) [inline doc/source/user/building.rst]: `libFLAME`?

**Review by rgommers — COMMENTED** [review]
Looks good, thanks @zerothi. I just suggested a few small tweaks.

File: doc/release/1.17.0-notes.rst (hunk 1 above)
Comment (zerothi) [reply doc/release/1.17.0-notes.rst]: how embarrassing, my eyes read timesort down in the text, hence I quickly *fixed* it... sorry for this, I'll fix.

File: doc/source/user/building.rst (hunk 4 above)
Comment (zerothi) [reply doc/source/user/building.rst]: yes, agreed

**Comment by zerothi** [comment]
@rgommers I can't see why the Windows Python36 fails, but Windows Python37 does not... It seems I don't have permission to re-run the test?

**Comment by rgommers** [comment]
Hmm, that looks unrelated but since it's in linalg I'd like to rerun the tests. Unfortunately I can't trigger a rerun either. @tylerjereddy could you trigger the rerun on Azure, and perhaps also add me (`ralf.gommers@gmail.com`) to the account?

**Comment by charris** [comment]
I've triggered the re-run.

**Comment by tylerjereddy** [comment]
Failure is still there for 32-bit Python 3.6 Windows, so maybe real. Various ways to debug--normally I'd set up Azure on a fork and play with the CI until I identify the specific test that crashes--maybe using a more verbose output.

@rgommers ok, you're an admin for NumPy Azure account now. Stephan Hoyer, Matti, and Chuck should also be admins.

**Comment by zerothi** [comment]
@tylerjereddy @rgommers I am still very puzzled by the apparent show of linalg error in *one* build.
I haven't changed any default build-options, only enabled more.

Is there anything I can do?

File: numpy/distutils/system_info.py [hunk 5]
```diff
@@ -2042,6 +2058,85 @@ def calc_info(self):
...
+        int main(int argc, const char *argv[])
+        {
+            zungqr_();
+            return 0;
+        }"""
+        src = os.path.join(tmpdir, 'source.c')
+        out = os.path.join(tmpdir, 'a.out')
+        # Add the additional "extra" arguments
+        extra_args = info.get('extra_link_args', [])
+        if sys.version_info < (3, 5) and sys.version_info > (3, 0) and c.compiler_type == "msvc":
```
Comment (rgommers) [inline numpy/distutils/system_info.py]: this can be removed, Python 3.5 is the lowest supported Python version

**Comment by rgommers** [comment]
Hmm the build logs for the failures have disappeared. I've triggered a re-run

File: numpy/distutils/system_info.py [hunk 6]
```diff
@@ -2042,6 +2058,85 @@ def calc_info(self):
...
+            try:
+                c.link_executable(obj, out, libraries=info['libraries'],
+                                  library_dirs=info['library_dirs'],
+                                  extra_postargs=extra_args)
+                return True
+            except distutils.ccompiler.LinkError:
+                return False
+        finally:
+            shutil.rmtree(tmpdir)
+        return False
```
Comment (tylerjereddy) [inline numpy/distutils/system_info.py]: The `return False` here is showing up as unreachable in the LGTM static analysis.

File: numpy/distutils/system_info.py (hunk 5 above)
Comment (zerothi) [reply numpy/distutils/system_info.py]: done

File: numpy/distutils/system_info.py (hunk 6 above)
Comment (zerothi) [reply numpy/distutils/system_info.py]: done

**Comment by rgommers** [comment]
All green, in it goes. Thanks @zerothi!

**Pull request merged**